### PR TITLE
Replace OC.Notification with toastify js

### PR DESCRIPTION
**Pull request by juliusknorr**
This is a PoC pull request for https://github.com/nextcloud/server/issues/13423 to make error messages that are sent though OC.Notification more obvious and visually appealing.

image

Feedback is welcome @nextcloud/designers 

Note that the current OC.Notification.show OC.Notification.showTemporary continue to work, but the toastify lib adds additional types (error,warning,success,info) to give a visual hint about the type of message.

ToDo as from https://github.com/nextcloud/server/issues/13423:
- [x] Every notification needs to be dismissable 
- [x] Move the toast to the top right

### CODE REVIEW

**Comment by rullzer** [comment]
So the design team has to decide of course. But I'm not 100% convinced having it on the bottom is the best. All our notifications etc are on the top so that is where people look. Maybe adding it just blow the header makes more sense?

**Comment by rullzer** [comment]
Of course visually this looks a lot better than the 1990 style we had before.

**Comment by stefan-niedermann** [comment]
I think it is hard to distinguish from the other content. Maybe use a contrast color (white on dark background on normal, and dark on white background on dark theme)?

**Comment by MorrisJobke** [comment]
> So the design team has to decide of course. But I'm not 100% convinced having it on the bottom is the best. All our notifications etc are on the top so that is where people look. Maybe adding it just blow the header makes more sense?

See the last open checkbox ;) It should be moved to the top right.

**Comment by juliushaertl** [comment]
> I think it is hard to distinguish from the other content. Maybe use a contrast color (white on dark background on normal, and dark on white background on dark theme)?

A full colored background is a bit to much I think, but I've increased the box shadow visibility a bit (see the updated screenshot in the first post)

> So the design team has to decide of course. But I'm not 100% convinced having it on the bottom is the best. All our notifications etc are on the top so that is where people look. Maybe adding it just blow the header makes more sense?

Done :wink:

**Comment by jancborchardt** [comment]
Top right is indeed much better, that’s where all the "meta / organizational" stuff is. :)

And it looks really really nice! Props @juliushaertl! The only thing I can find is that the x icon is not our `.icon-close` icon – any reason for that?

**Comment by juliushaertl** [comment]
> And it looks really really nice! Props @juliushaertl! The only thing I can find is that the x icon is not our .icon-close icon – any reason for that?

No reason, I can adjust that as well.

**Comment by juliushaertl** [comment]
@jancborchardt Thanks for the hint, pushed with the proper close icon 
image

File: core/src/OC/notification.js [hunk 1]
```diff
@@ -205,21 +163,9 @@ export default {
...
-		_.defaults(options, defaults);
-
-		var $row;
-		if (options.isHTML) {
-			$row = this.showHtml(text, options);
-		} else {
-			$row = this.show(text, options);
-		}
-		return $row;
+		options = options || {}
```
Comment (skjnldsv) [inline core/src/OC/notification.js]: Shall we deprecate in favour of OCP.Toast?

**Review by skjnldsv — CHANGES_REQUESTED** [review]
Having the mouse over the notification should prevent it from disappearing I think. If you're reading something long and it disappear, you're confused :)

File: core/src/OC/notification.js (hunk 1 above)
Comment (juliushaertl) [reply core/src/OC/notification.js]: Ah yes, good point. :+1: Notification was always a bit confusing with the notifications app :wink:

**Comment by juliushaertl** [comment]
> Having the mouse over the notification should prevent it from disappearing I think. If you're reading something long and it disappear, you're confused :)

I'll see, but we probably need to fix that upstream.

**Comment by BrookeDot** [comment]
> Having the mouse over the notification should prevent it from disappearing I think. If you're reading something long and it disappear, you're confused :)

I agree, I struggle from a bit of dyslexia and often find myself highlighting text to help me read it and sometimes reading takes a bit longer than "average". Having a notification disappear while reading it can be quite frustrating. Especially if there's not an easy way to get it back. 

> I'll see, but we probably need to fix that upstream.

I have opened up https://github.com/apvarun/toastify-js/issues/21  upstream so we'll see if this feature can get added. If not, perhaps I can recommend not setting the timeout at all.

**Comment by skjnldsv** [comment]
@BrookeDot Thanks a lot for that! :rocket:

**Comment by skjnldsv** [comment]
> I have opened up [apvarun/toastify-js#21](https://github.com/apvarun/toastify-js/issues/21) upstream so we'll see if this feature can get added. If not, perhaps I can recommend not setting the timeout at all.

Opened PR: https://github.com/apvarun/toastify-js/pull/22

**Comment by karlitschek** [comment]
I like it. Looks more modern

**Comment by juliushaertl** [comment]
I deprecated the old OC.Notification API, should we get this in for now? We can update the dependency later once https://github.com/apvarun/toastify-js/pull/22 is merged.

**Comment by rullzer** [comment]
@juliushaertl yes lets get it in and then update later :+1: else we keep rebasing etc.

**Comment by skjnldsv** [comment]
Agreed!
tumblr_mq00lsWsTd1stru2ro6_250

**Review by skjnldsv — APPROVED** [review]

**Comment by MorrisJobke** [comment]
Conflicts 🙈

**Comment by juliushaertl** [comment]
Did another rebase and removed to fixup from the commit messages.

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
@juliushaertl Mind to fix the conflicts?

**Comment by MorrisJobke** [comment]
> @juliushaertl Mind to fix the conflicts?

Done.

**Comment by MorrisJobke** [comment]
JSUnit failures :/

**Comment by rullzer** [comment]
Need a rebase and the jsunit fixed... then we can get this in

**Comment by juliushaertl** [comment]
Yep I'll look into the failing tests.

**Comment by juliushaertl** [comment]
While looking at the tests, we still lack of proper backward compatibility for the `OC.Notification.hide` method. I created https://github.com/apvarun/toastify-js/pull/26 to enhance the library, so that we can have permanent toasts that can be closed as it is currently done in the OC.Notifications implementation.

**Comment by MorrisJobke** [comment]
> While looking at the tests, we still lack of proper backward compatibility for the `OC.Notification.hide` method. I created [apvarun/toastify-js#26](https://github.com/apvarun/toastify-js/pull/26) to enhance the library, so that we can have permanent toasts that can be closed as it is currently done in the OC.Notifications implementation.

Do we really need that this often?

**Comment by MorrisJobke** [comment]
> Do we really need that this often?

In most cases it is called to avoid that an old notification is hiding the one that should be added.

**Comment by juliushaertl** [comment]
>> Do we really need that this often?
> In most cases it is called to avoid that an old notification is hiding the one that should be added.

I would also be happy with just ignoring the hide method and have a default timeout of 30s for notifications that were called without a timeout though the legacy API. 
Anything that speaks against this @nextcloud/designers ?

**Comment by jancborchardt** [comment]
> I would also be happy with just ignoring the hide method and have a default timeout of 30s for notifications that were called without a timeout though the legacy API.
>  Anything that speaks against this @nextcloud/designers ?

Just to clarify: This issue does not happen when using this new notification API?

Cause a defaut timeout for notifications has a big issue in that people don’t necessarily check the screen or tab all the time, and they might miss a notification. Also, do these notifications have a close button so you can close them _before_ the 30s?

**Comment by juliushaertl** [comment]
> Just to clarify: This issue does not happen when using this new notification API?

The new API has a default timeout as I would consider those notifications to be more temporary than permanent.

> Cause a defaut timeout for notifications has a big issue in that people don’t necessarily check the screen or tab all the time, and they might miss a notification. Also, do these notifications have a close button so you can close them before the 30s?

I have now also rebuild the old OC.Notifiation.hide() method, so the behavior on the old API should be complete. The only thing that has changed is the fallback if no notification to hide was provided. In that case we do nothing now, as there was a warning logged for quite some time and just removing the first notification doesn't seem to be a good fallback to me.

Let's see if the tests are happy.

**Comment by juliushaertl** [comment]
:tada: Tests are happy again. @skjnldsv @ChristophWurst Mind to have another look?

**Comment by skjnldsv** [comment]
Looks all good to me :D
Rebase, sign and squash! :rocket:

**Comment by juliushaertl** [comment]
> Rebase, sign and squash!

Done :ship:

**Comment by danxuliu** [comment]
@juliushaertl There is a mismatch in the options received by OC.Notification and OCP.Toast: [OC.Notification uses isHTML](https://github.com/nextcloud/server/blob/2d0337332d64065c5bf36dfb775c48976c0f730d/core/src/OC/notification.js#L142), while [OCP.Toast uses showHtml](https://github.com/nextcloud/server/blob/6d848b545df054b994ca056a90c72661e4ef6e95/core/src/OCP/toast.js#L59). Either [`OCP.Toast.message` is adjusted](https://github.com/nextcloud/server/blob/6d848b545df054b994ca056a90c72661e4ef6e95/core/src/OCP/toast.js#L59-L64) to use `isHTML` too, or [`OC.Notification.showTemporary` is adjusted](https://github.com/nextcloud/server/blob/2d0337332d64065c5bf36dfb775c48976c0f730d/core/src/OC/notification.js#L147-L149) to set `options.showHtml` based on `options.isHTML` (and also the unit tests to reflect that [`OC.Toast.message` expects `showHtml` instead of `isHTML`](https://github.com/nextcloud/server/blob/6d848b545df054b994ca056a90c72661e4ef6e95/core/js/tests/specs/coreSpec.js#L941)).

**Comment by LinneyS** [comment]
If you scroll down the page, the message is not visible under the header

**Comment by ChristophWurst** [comment]
> If you scroll down the page, the message is not visible under the header

Please open a new ticket.

**Comment by skjnldsv** [comment]
Should be fixed already https://github.com/nextcloud/server/pull/17110

**Pull request merged**